From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Manish Bhardwaj <mbhardwaj@nvidia.com>
Date: Wed, 24 Sep 2025 03:55:43 +0000
Subject: [PATCH 110/243] NVIDIA: SAUCE: tegra_hv: Update interrupt property
 handling

Enhanced interrupt property handling by dynamically reading
"#interrupt-cells" from device tree, ensuring correct configuration
for different hardware setups.

Bug 5520586

Change-Id: Ia7bddf02553b9d8845c4f120413369da7ad7379a
Signed-off-by: Manish Bhardwaj <mbhardwaj@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3457115
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
Reviewed-by: Sandeep Trasi <strasi@nvidia.com>
Reviewed-by: svcacv <svcacv@nvidia.com>
Reviewed-by: Suresh Mangipudi <smangipudi@nvidia.com>
(cherry picked from commit 68ae606c6b62fb95cbabf494bcd55470b842b724)

diff --git a/drivers/virt/tegra/tegra_hv.c b/drivers/virt/tegra/tegra_hv.c
index 2f38cadb4a7d0..c2ce088cbe964 100644
--- a/drivers/virt/tegra/tegra_hv.c
+++ b/drivers/virt/tegra/tegra_hv.c
@@ -590,7 +590,8 @@ static struct ivc_info_page *validate_ivc_info_page(void __iomem *mapped_mem, si
 
 static int tegra_hv_setup(struct tegra_hv_data *hvd)
 {
-	const uint32_t intr_property_size = 3u;
+	uint32_t intr_property_size;
+	struct device_node *p;
 	void __iomem *mapped_mem;
 	uint64_t info_page;
 	uint32_t i, result;
@@ -604,6 +605,20 @@ static int tegra_hv_setup(struct tegra_hv_data *hvd)
 		return -ENODEV;
 	}
 
+	p = of_irq_find_parent(hvd->dev);
+	if (p) {
+		if (of_property_read_u32(p, "#interrupt-cells", &intr_property_size)) {
+			ERR("Get interrupt-cells failed\n");
+			intr_property_size = 3;
+		}
+		of_node_put(p);
+	} else {
+		ERR("Null device parent node\n");
+		intr_property_size = 3;
+	}
+
+	INFO("Get interrupt-cells=%d\n", intr_property_size);
+
 	ret = hyp_read_gid(&hvd->guestid);
 	if (ret != 0) {
 		ERR("Failed to read guest id\n");
@@ -757,6 +772,14 @@ static int tegra_hv_setup(struct tegra_hv_data *hvd)
 			(__force uint32_t)cpu_to_be32(qd->irq - 32); /* Id in SPI namespace */
 		/* 0x1 == low-to-high edge */
 		hvd->interrupts_arr[result + 2] = (__force uint32_t)cpu_to_be32(0x1);
+		/* The 4th cell is a phandle to a node describing a set of CPUs this
+		 * interrupt is affine to. The interrupt must be a PPI, and the node
+		 * pointed must be a subnode of the "ppi-partitions" subnode. For
+		 * interrupt types other than PPI or PPIs that are not partitioned,
+		 * this cell must be zero. See the "ppi-partitions" node description below.
+		 */
+		if (intr_property_size > 3)
+			hvd->interrupts_arr[result + 3] = (__force uint32_t)cpu_to_be32(0);
 	}
 
 	interrupts_prop.length =
-- 
2.43.0

